### ISSUE TITLE
Training loss becomes negative (self-critical with topdown) 

### ISSUE THREAD

**Issue by KahHwa** [ruotianluo/self-critical.pytorch#79]
I notice the training loss becoming negative after 6000 iterations training with self-critical. Is this normal?

**Comment by ruotianluo** [comment]
That means average reward turns to positive.

**Comment by KahHwa** [comment]
So, it means that the model is sampling almost every word correctly? In the context of image captioning, is over-fitting a good or a bad thing? Because the cider, bleu scores are still improving.

**Comment by ruotianluo** [comment]
It means sampling is getting better results than greedy decoding.

**Comment by KahHwa** [comment]
What do you mean? When i train the model, my beam-size is set to 1. Isn't this greedy decoding ??

**Comment by ruotianluo** [comment]
the reward is calculated by substracting cider score of sampling result by cider score of greedy decoding result.
If reward is greater than zero, it means sampling result is better than greedy decoding result.

**Comment by Daibingh** [comment]
> That means average reward turns to positive.

Is that correct?  I find the following code [https://github.com/ruotianluo/self-critical.pytorch/blob/master/misc/utils.py#L115](url)

`class RewardCriterion(nn.Module):
    def __init__(self):
        super(RewardCriterion, self).__init__()

    def forward(self, input, seq, reward):
        input = input.gather(2, seq.unsqueeze(2)).squeeze(2)
        
        input = input.reshape(-1)
        reward = reward.reshape(-1)
        mask = (seq>0).float()
        mask = torch.cat([mask.new(mask.size(0), 1).fill_(1), mask[:, :-1]], 1).reshape(-1)
        output = - input * reward * mask
        output = torch.sum(output) / torch.sum(mask)

        return output`

the reward loss = - logprob * reward * mask, the logprobs are always negative, so the sign of loss must equal to the reward. why  do you say that "training loss becoming negative means s average reward turns to positive."

**Comment by ruotianluo** [comment]
You are right. I was wrong.
